File: src/main/java/fi/quanfoxes/assembler/references/MemoryReference.java
package fi.quanfoxes.assembler.references;

import fi.quanfoxes.assembler.*;

public class MemoryReference extends Reference {
    private Register register;
    private int alignment;

    public MemoryReference(Register register, int alignment, int size) {
        super(Size.get(size));

        this.register = register;
        this.alignment = alignment;
    }

    public String getContent() {
        if (alignment > 0) {
            return String.format("%s+%d", register, alignment);
        }
        else if (alignment < 0) {
            return String.format("%s%d", register, alignment);
        }
        else {
            return String.format("%s", register);
        }
    }

    @Override
    public String use() {
        return String.format("[%s]", getContent());
    }

    @Override
    public boolean isComplex() {
        return true;
    }

    public static MemoryReference local(Unit unit, int alignment, int size) {
        return new MemoryReference(unit.ebp, -alignment - size, size);
    }

    public static MemoryReference parameter(Unit unit, int alignment, int size) {
        return new MemoryReference(unit.ebp, alignment + 8, size);
    }

    public static MemoryReference member(Register register, int alignment, int size) {
        return new MemoryReference(register, alignment, size);
    }

    public LocationType getType() {
        return LocationType.MEMORY;
    }
}

File: Zigzag/Assembler/References/MemoryReference.cs
public class MemoryReference : Reference
{
	private Register Register { get; set; }
	private int Alignment { get; set; }

	public MemoryReference(Register register, int alignment, int size) : this(register, alignment, Size.Get(size)) { }

	public MemoryReference(Register register, int alignment, Size size) : base(size)
	{
		Register = register;
		Alignment = alignment;
	}

	public string GetContent()
	{
		if (Alignment > 0)
		{
			return $"{Register}+{Alignment}";
		}
		else if (Alignment < 0)
		{
			return $"{Register}{Alignment}";
		}
		else
		{
			return $"{Register}";
		}
	}

	public override string Use(Size size)
	{
		return $"{size} [{GetContent()}]";
	}

	public override bool IsComplex()
	{
		return true;
	}
	public override LocationType GetType()
	{
		return LocationType.MEMORY;
	}

	public static MemoryReference Local(Unit unit, int alignment, int size)
	{
		return new MemoryReference(unit.EBP, -alignment - size, size);
	}

	public static MemoryReference Parameter(Unit unit, int alignment, int size)
	{
		return new MemoryReference(unit.EBP, alignment + 8, size);
	}

	public static MemoryReference Member(Register register, int alignment, int size)
	{
		return new MemoryReference(register, alignment, size);
	}
}